tests/test_socket_server.py

```diff
@@ -0,0 +1,180 @@
+        """Write the data appended with a tcpip.TERMINATOR string.
+
+        Parameters
+        ----------
+        data:
+            The data to write.
+        """
+        st = json.dumps({**data})
+        assert self.writer is not None
+        self.writer.write(st.encode() + tcpip.TERMINATOR)
+        await self.writer.drain()
+
+    async def test_disconnect(self) -> None:
+        self.assertTrue(self.srv.connected)
+        await self.write(command=common.Command.DISCONNECT, parameters={})
+        # Give time to the socket server to clean up internal state and exit.
+        await asyncio.sleep(0.5)
+        self.assertFalse(self.srv.connected)
+
+    async def test_exit(self) -> None:
+        self.assertTrue(self.srv.connected)
+        await self.write(command=common.Command.EXIT, parameters={})
+        # Give time to the socket server to clean up internal state and exit.
+        await asyncio.sleep(0.5)
+        self.assertFalse(self.srv.connected)
+
+    def check_temperature_reply(
+        self, reply: typing.List[typing.Union[str, float]]
+    ) -> None:
+        device_name = reply[0]
+        time = reply[1]
+        response_code = reply[2]
+        resp = reply[3:]
+
+        self.assertEqual(self.name, device_name)
+        self.assertGreater(time, 0)
+        if self.in_error_state:
+            self.assertEqual(common.ResponseCode.DEVICE_READ_ERROR, response_code)
+        else:
+            self.assertEqual(common.ResponseCode.OK, response_code)
+        self.assertEqual(len(resp), self.num_channels)
+        for i in range(0, self.num_channels):
+            self.assertLessEqual(common.MockTemperatureConfig.min, resp[i])
+            self.assertLessEqual(resp[i], common.MockTemperatureConfig.max)
+
+    async def check_server_test(self) -> None:
+        """Test a full command sequence of the SocketServer.
+
+        The sequence is
+            - configure
+            - start
+            - read telemetry
+            - stop
+            - disconnect
+            - exit
+        """
+        configuration = {
+            common.Key.DEVICES: [
+                {
+                    common.Key.NAME: self.name,
+                    common.Key.CHANNELS: self.num_channels,
+                    common.Key.DEVICE_TYPE: common.DeviceType.FTDI,
+                    common.Key.FTDI_ID: "ABC",
+                    common.Key.SENSOR_TYPE: common.SensorType.TEMPERATURE,
+                }
+            ]
+        }
+        await self.write(
+            command=common.Command.CONFIGURE,
+            parameters={common.Key.CONFIGURATION: configuration},
+        )
+        data = await self.read()
+        self.assertEqual(common.ResponseCode.OK, data[common.Key.RESPONSE])
+        await self.write(command=common.Command.START, parameters={})
+        data = await self.read()
+        self.assertEqual(common.ResponseCode.OK, data[common.Key.RESPONSE])
+
+        self.reply = await self.read()
```

Comment:
I strongly suggest not setting `self.reply` here.

It's an instance of my comment about self.num_channels with the added problem that it's not used outside of this method so it seems to be pure clutter.

If you want to examine the replies afterwards you could return an array of replies (or if that is not practical,  consider having setUp create an array for saving replies and then fill it as appropriate).

Reply:
I agree it is pure clutter and I made it a local variable.